Memoize call expressions in optional chains in loose mode

oliverdunk opened the pull request:
<!--
Before making a PR, please read our contributing guidelines
https://github.com/babel/babel/blob/master/CONTRIBUTING.md

Please note that the Babel Team requires two approvals before merging most PRs.

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/) followed by the ticket number fixed by the PR. (it should be underlined in the preview if done correctly)

If you are making a change that should have a docs update: submit another PR to https://github.com/babel/website
-->

| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | Fixes #11084  <!-- remove the (`) quotes and write "Fixes" before the number to link the issues -->
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  | No
| Minor: New Feature?      | No
| Tests Added + Pass?      | Yes
| Documentation PR Link    | <!-- If only readme change, add `[skip ci]` to your commits -->
| Any Dependency Changes?  | No
| License                  | MIT

Previously, in loose mode, we weren't memoizing optional call expressions. This resulted in code like `a.get(b)?.();` being compiled to `a.get(b) == null ? void 0 : a.get(b)();`, which repeats a function call.

This MR changes the behaviour so that memoization is only skipped when no work is done by the call's callee.

jridgewell commented on packages/babel-plugin-proposal-optional-chaining/src/index.js:
```diff
@@ -50,9 +59,10 @@ export default declare((api, options) => {
           let ref;
           let check;
-          if (loose && isCall) {
+          if (loose && isCall && isSimpleMemberExpression(node.callee)) {
```
I think this can be as simple as `!t.isCallExpression(chain)` and still get the same behavior.

Note, `chain` is just a local variable for `node.callee`.

jridgewell commented on packages/babel-plugin-proposal-optional-chaining/src/index.js:
```diff
@@ -50,9 +59,10 @@ export default declare((api, options) => {
           let ref;
           let check;
-          if (loose && isCall) {
+          if (loose && isCall && isSimpleMemberExpression(node.callee)) {
```
Nit:
```suggestion
          if (loose && isCall && isSimpleMemberExpression(chain)) {
```

oliverdunk replied:
Easy change, done!

oliverdunk commented:
@jridgewell Please ignore that request for review. I need to make sure we're probably setting the call context.

oliverdunk commented:
@jridgewell Found and fixed my bug! It would be great if you could take a look :)

jridgewell commented on packages/babel-plugin-proposal-optional-chaining/src/index.js:
```diff
@@ -72,12 +83,12 @@ export default declare((api, options) => {
 
           // Ensure call expressions have the proper `this`
           // `foo.bar()` has context `foo`.
-          if (isCall && t.isMemberExpression(chain)) {
```
We'll still need this.

oliverdunk replied:
Ah, I was wondering why the chain.object wasn't needed before. Do you think isMemberExpression is a bit too specific? Shouldn't we allow optional member expressions too?

jridgewell replied:
Any optional members will have been converted to regular members by https://github.com/babel/babel/blob/d4d6ec8f2b8ebfb32381a1e8dcadbadec3ae5736/packages/babel-plugin-proposal-optional-chaining/src/index.js#L41

jridgewell commented on packages/babel-plugin-proposal-optional-chaining/src/index.js:
```diff
@@ -72,12 +83,12 @@ export default declare((api, options) => {
               // `a.?b.?()` translates roughly to `_a.b != null && _a.b()`
               node.callee = chain;
-            } else {
+            } else if (chain.object) {
```
With the above change, we'll be able to revert this.

oliverdunk commented:
@jridgewell I've addressed your feedback again. Really appreciate all of the help! It's early days when it comes to my understanding of Babel - but you and Nico have been incredibly welcoming.

nicolo-ribaudo commented:
Thanks Oliver!
